### PR TITLE
Database-based storage

### PR DESCRIPTION
**Pull request by George3d6**
## Why
* We need a centralized state containing datasources, configs and predicotrs that can be used by multiple mindsdb
* We need this centralized state to be easy to synchronize

## How
* In short, by moving the "logic" data layer of mindsdb to a database (implemented via the sqlalchemy ORM)

## Extra
* Removed installer code
* Removed a bunch of already dead code
* Removed code that was no longer necessary

### CODE REVIEW

**Comment by StpMax** [comment]
Missed value in 'datasource' column on a 'predictors' screen. With staging branch that values exists (equal to csv file name).
image

**Comment by StpMax** [comment]
Here is missed DatabaseWrapper: https://github.com/mindsdb/mindsdb/blob/distribution_2/mindsdb/api/http/namespaces/config.py#L77

But if add `dbw = DatabaseWrapper(ca.config_obj)`, then each click on 'test database connection' at 'Database Integration' scree n will create new integration.

File: mindsdb/api/http/namespaces/config.py [hunk 1]
```diff
@@ -69,11 +69,6 @@ def put(self, name):
                 del params['enabled']
             ca.config_obj.add_db_integration(name, params)
 
-            mdb = ca.mindsdb_native
-            cst = ca.custom_models
-            model_data_arr = get_all_models_meta_data(mdb, cst)
-            dbw = DatabaseWrapper(ca.config_obj)
```
Comment (StpMax) [inline mindsdb/api/http/namespaces/config.py:75]: dbw need for code below

**Review by StpMax — CHANGES_REQUESTED** [review]

**Comment by StpMax** [comment]
Training new predictor through GUI creates many new integrations. Probably it same as issue above.

File: mindsdb/interfaces/database/database.py [hunk 2]
```diff
@@ -48,17 +49,24 @@ def _get_integrations(self):
...
+            else:
+                model_arr = [model]
+            try:
+                for integration in it:
+                    register = True
+                    if setup:
+                        register = self._setup_integration(integration)
+                    if register:
+                        if integration.check_connection():
+                            integration.register_predictors(model_arr)
```
Comment (StpMax) [inline mindsdb/interfaces/database/database.py:64]: after predictor training finished, that line called 2 times for same integration. That leads to this output:
```
--- Logging error ---
Traceback (most recent call last):
  File "/home/maxs/dev/mdb/venv_new/sources/mindsdb/mindsdb/interfaces/database/database.py", line 65, in register_predictors
    integration.register_predictors(model_arr)
  File "/home/maxs/dev/mdb/venv_new/sources/mindsdb/mindsdb/integrations/postgres/postgres.py", line 140, in register_predictors
    self._query(q)
  File "/home/maxs/dev/mdb/venv_new/sources/mindsdb/mindsdb/integrations/postgres/postgres.py", line 52, in _query
    cur.execute(query)
  File "/home/maxs/dev/mdb/venv_new/lib/python3.6/site-packages/pg8000/core.py", line 865, in execute
    self._c.execute(self, operation, args)
  File "/home/maxs/dev/mdb/venv_new/lib/python3.6/site-packages/pg8000/core.py", line 1925, in execute
    self.handle_messages(cursor)
  File "/home/maxs/dev/mdb/venv_new/lib/python3.6/site-packages/pg8000/core.py", line 1992, in handle_messages
    raise self.error
pg8000.core.ProgrammingError: {'S': 'ERROR', 'V': 'ERROR', 'C': '42P07', 'M': 'relation "carp_p4" already exists', 'F': 'heap.c', 'L': '1155', 'R': 'heap_create_with_catalog'}

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/logging/__init__.py", line 994, in emit
    msg = self.format(record)
  File "/usr/lib/python3.6/logging/__init__.py", line 840, in format
    return fmt.format(record)
  File "/usr/lib/python3.6/logging/__init__.py", line 577, in format
    record.message = record.getMessage()
  File "/usr/lib/python3.6/logging/__init__.py", line 338, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.6/multiprocessing/spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "/usr/lib/python3.6/multiprocessing/spawn.py", line 118, in _main
    return self._bootstrap()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/home/maxs/dev/mdb/venv_new/sources/mindsdb/mindsdb/interfaces/native/predictor_process.py", line 45, in run
    state.update_predictor(name=name, status=status, original_path=None, data=json.dumps(analysis))
  File "/home/maxs/dev/mdb/venv_new/sources/mindsdb/mindsdb/interfaces/state/state.py", line 73, in update_predictor
    }], False)
  File "/home/maxs/dev/mdb/venv_new/sources/mindsdb/mindsdb/interfaces/database/database.py", line 70, in register_predictors
    logger.error('Cloud not regier model: ', e)
Message: 'Cloud not regier model: '
Arguments: (ProgrammingError({'S': 'ERROR', 'V': 'ERROR', 'C': '42P07', 'M': 'relation "carp_p4" already exists', 'F': 'heap.c', 'L': '1155', 'R': 'heap_create_with_catalog'},),)
```

File: mindsdb/interfaces/native/mindsdb.py [hunk 3]
```diff
@@ -90,21 +110,27 @@ def get_models(self, status='any'):
...
         F.rename_model(name, new_name)
-        self.dbw.register_predictors(self.get_model_data(new_name), setup=False)
+        self.state.rename_predictor(name, new_name)
 
     def load_model(self, fpath):
+        # self.state.make_predictor(name, None, to_predict) <--- fix
         F.import_model(model_archive_path=fpath)
-        # @TODO How do we figure out the name here ?
-        # dbw.register_predictors(...)
+        self.state.make_predictor(name, None, to_predict, False)
```
Comment (StpMax) [inline mindsdb/interfaces/native/mindsdb.py:127]:
```
ERROR:mindsdb.api.http.initialize:Exception on /api/predictors/upload [POST]
Traceback (most recent call last):
  File "/home/maxs/dev/mdb/venv_new/sources/flask/src/flask/app.py", line 1949, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/maxs/dev/mdb/venv_new/sources/flask/src/flask/app.py", line 1935, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/maxs/dev/mdb/venv_new/sources/flask-restx/flask_restx/api.py", line 375, in wrapper
    resp = resource(*args, **kwargs)
  File "/home/maxs/dev/mdb/venv_new/sources/flask/src/flask/views.py", line 89, in view
    return self.dispatch_request(*args, **kwargs)
  File "/home/maxs/dev/mdb/venv_new/sources/flask-restx/flask_restx/resource.py", line 44, in dispatch_request
    resp = meth(*args, **kwargs)
  File "/home/maxs/dev/mdb/venv_new/sources/mindsdb/mindsdb/api/http/namespaces/predictor.py", line 276, in post
    ca.mindsdb_native.load_model(fpath)
  File "/home/maxs/dev/mdb/venv_new/sources/mindsdb/mindsdb/interfaces/native/mindsdb.py", line 127, in load_model
    self.state.make_predictor(name, None, to_predict, False)
NameError: name 'name' is not defined
```

Comment (StpMax) [reply mindsdb/interfaces/native/mindsdb.py:127]: ant to_predict

**Comment by George3d6** [comment]
> Here is missed DatabaseWrapper: https://github.com/mindsdb/mindsdb/blob/distribution_2/mindsdb/api/http/namespaces/config.py#L77
> 
> But if add `dbw = DatabaseWrapper(ca.config_obj)`, then each click on 'test database connection' at 'Database Integration' scree n will create new integration.

**Comment by George3d6** [comment]
> Missed value in 'datasource' column on a 'predictors' screen. With staging branch that values exists (equal to csv file name).
> image

The predictor object now carrier around a `datasource_id` we can merge on to get the actual datasource name, *however* datasources aren't passed around in an ideal way at the moment, in the future they should be passed around as just the metadata (including the `id`) and only instantiated (turned into a dataframe or a native datasource) when learn is called.

Alas, I didn't have the time to get this in and the PR was getting too big, so I'm reserving this change for later, hopefully before next week.

File: mindsdb/interfaces/database/database.py (hunk 2 above)
Comment (ZoranPandovski) [reply mindsdb/interfaces/database/database.py:64]: Same for me

File: mindsdb/interfaces/native/predictor_process.py [hunk 4]
```diff
@@ -19,27 +21,32 @@ def run(self):
...
+                mdb.learn(
+                    from_data=data_source,
+                    to_predict=to_predict,
+                    **kwargs
+                )
+
+                analysis = mindsdb_native.F.get_model_data(name)
+                status = analysis['status']
+
+                state.update_predictor(name=name, status=status, original_path=None, data=json.dumps(analysis))
```
Comment (StpMax) [inline mindsdb/interfaces/native/predictor_process.py:45]:
```
Traceback (most recent call last):
  File "/home/maxs/dev/mdb/venv_new/sources/mindsdb/mindsdb/interfaces/native/predictor_process.py", line 45, in run
    state.update_predictor(name=name, status=status, original_path=None, data=json.dumps(analysis))
TypeError: update_predictor() missing 1 required positional argument: 'ds_name'
```

**Comment by George3d6** [comment]
Killing this, way too bugy.

**Pull request closed without merging**